ddtrace/bootstrap/sitecustomize.py

```diff
@@ -111,7 +133,17 @@ def update_patched_modules():
         update_patched_modules()
         from ddtrace import patch_all
 
+        # We need to clean up after we have imported everything we need from
+        # ddtrace, but before we register the patch-on-import hooks for the
+        # integrations.
+        cleanup_loaded_modules()
+
         patch_all(**EXTRA_PATCHED_MODULES)
+    else:
+        cleanup_loaded_modules()
```

Comment:
After this point, when the traced application starts, it will import its own copy of the needed modules. If `gevent.monkey.patch_all()` is called, it will patch the newly loaded modules, but won't affect the modules that have been loaded during the initialisation of our library.

Reply:
Based on early conversations about this approach, I was looking for a place where "two copies" of a module were held, one for use by `ddtrace` internals and the other for use by user code.

Here's the sequence of events after this change as I currently understand it. I'd appreciate if you could tell me where I'm wrong @P403n1x87.

1. ddtracepy runs `import threading`. `'threading' in sys.modules`, `threading` names bound into a namespace internal to ddtracepy
2. ddtracepy starts a thread
3. ddtracepy calls `cleanup_loaded_modules`. `'threading' not in sys.modules`, `threading` names still bound in internal namespace
4. user code runs `import gevent.monkey; gevent.monkey.patch_all()`. does this affect the running thread in any way?
5. user code runs `import threading`. because `'threading' not in sys.modules`, the module is fetched from disk and names are bound into a user-code namespace
6. from this point on, names from the `threading` module in both user code and ddtracepy code reference the copy of the module loaded in user code (and monkeypatched by gevent). objects (like threads) instantiated at step 2 continue to exist as instances of names from the module loaded in step 1.

In this sequence, "two copies" of the `threading` module aren't bound to different namespaces at the same moment. Instead, objects created via the access of names from `threading` existing at the same moment could be from either the pre- or post-monkeypatch names. That implies that our library actually needs to create all such objects that it will ever need before cleaning up its copy of `threading`.

Reply:
> 1. ddtracepy runs `import threading`. `'threading' in sys.modules`, `threading` names bound into a namespace internal to ddtracepy

Every module in ddtrace that imports the `threading` module will have a local reference to it that is set by the import statement.

> 2. ddtracepy starts a thread

and will use the `Thread` class from the vanilla `threading` module from the stdlib.

> 3. ddtracepy calls `cleanup_loaded_modules`. `'threading' not in sys.modules`, `threading` names still bound in internal namespace

Yep, we still hold a reference to the originally imported module everywhere it was `import`ed.

> 4. user code runs `import gevent.monkey; gevent.monkey.patch_all()`. does this affect the running thread in any way?

No. The thread was already created, and the `Thread` class used for it has access to only the original "stuff". When `patch_all` is called, a new copy of the `threading` module will be created, and the one referenced by `ddtrace` will be invisible to `gevent`. 

> 5. user code runs `import threading`. because `'threading' not in sys.modules`, the module is fetched from disk and names are bound into a user-code namespace

If `patch_all` was called, then `threading` would have been imported by `gevent`, so see comment to 4.

> 6. from this point on, names from the `threading` module in both user code and ddtracepy code reference the copy of the module loaded in user code (and monkeypatched by gevent). objects (like threads) instantiated at step 2 continue to exist as instances of names from the module loaded in step 1.

I'm not sure I follow entirely. At this point, ddtrace still has references to the first, unpatched copy of `threading`, whereas the application only sees the new, `gevent`-patched copy of the `threading` module. The same holds for all the modules imported by the `threading` module.

> In this sequence, "two copies" of the `threading` module aren't bound to different namespaces at the same moment.

Indeed they are, in the sense described above.

> Instead, objects created via the access of names from `threading` existing at the same moment could be from either the pre- or post-monkeypatch names. That implies that our library actually needs to create all such objects that it will ever need before cleaning up its copy of `threading`.

The objects ddtrace needs are created on `import`. The only modules we load after our initialisation are those that feature in integrations, because those will be "invoked" on import via the import hooks.

Reply:
Thanks so much for spelling this out @P403n1x87. Seems like the main detail I was misunderstanding was that the original references to `threading` continue to point to the same thing throughout the lifetime of the ddtrace code that imported them. If you know of any source where I can read more about how python's importing machinery works at this low level, please do recommend it.

Reply:
To my understanding, imports in contrib are subject to gevent patching.

Reply:
Since the major failure modes we've been seeing from gevent patching are related to forking/threading, I'm guessing gevent patching the contrib imports are not the problem. Is that right?